File: android-overrides/SensorProbePlugin.kt
// Capacitor plugin: stream every native-only ambient sensor the web
// layer can't reach — barometer (TYPE_PRESSURE), raw magnetometer
// (TYPE_MAGNETIC_FIELD), ambient light (TYPE_LIGHT), proximity
// (TYPE_PROXIMITY), battery level/state, thermal status, Battery Saver —
// for the Extras "Sensors" dashboard (static/extras-sensors.js). The
// iOS mirror is ios-overrides/SensorProbePlugin.swift; both register as
// jsName "SensorProbe" with the same contract so the JS side needs no
// platform branches:
//
//   getInfo() -> { barometer, magnetometer, light, proximity, thermal }
//   start()  -> begins 4 Hz 'reading' events (notifyListeners) with the
//        current snapshot: { pressureHPa?, magX/Y/Z?, lux?,
//        proximityCm?/proximityNear?, batteryPct?, batteryCharging?,
//        thermal?, lowPower }
//   stop()   -> tears the sensors down (the JS side calls this whenever
//        the dashboard is hidden, so nothing streams in the background;
//        handleOnPause is a native safety net for the same)
//
// Mapping notes vs the iOS half:
//   - relAltM (CMAltimeter's session altitude delta) has no Android
//     analogue and is simply absent — the JS row shows an em-dash.
//   - Thermal statuses are folded onto the iOS vocabulary so the JS
//     shows one scale: NONE->nominal, LIGHT/MODERATE->fair,
//     SEVERE->serious, CRITICAL/EMERGENCY/SHUTDOWN->critical.
//     currentThermalStatus needs API 29+; omitted below that.
//   - proximityNear uses the value < maximumRange convention (most
//     hardware reports a binary far==max / near==0).
//
// None of these sensors need a manifest permission. Registered from
// MainActivity's onCreate via registerPlugin(SensorProbePlugin.class)
// (a line the android-build.yml workflow injects); this file itself is
// picked up by the workflow's android-overrides/*.kt copy step.

package org.phylliidaassets.creaturecollect

import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.hardware.Sensor
import android.hardware.SensorEvent
import android.hardware.SensorEventListener
import android.hardware.SensorManager
import android.os.BatteryManager
import android.os.Build
import android.os.Handler
import android.os.Looper
import android.os.PowerManager
import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin

@CapacitorPlugin(name = "SensorProbe")
class SensorProbePlugin : Plugin(), SensorEventListener {

    private var running = false
    private var pressureHPa: Float? = null
    private var lux: Float? = null
    private var proximityCm: Float? = null
    private var proximityMax = 5f
    private var mag: FloatArray? = null
    private val handler = Handler(Looper.getMainLooper())
    private val emitter = object : Runnable {
        override fun run() {
            if (!running) return
            emit()
            handler.postDelayed(this, 250)
        }
    }

    private fun sensorManager(): SensorManager =
        context.getSystemService(Context.SENSOR_SERVICE) as SensorManager

    @PluginMethod
    fun getInfo(call: PluginCall) {
        val sm = sensorManager()
        val out = JSObject()
        out.put("barometer", sm.getDefaultSensor(Sensor.TYPE_PRESSURE) != null)
        out.put("magnetometer", sm.getDefaultSensor(Sensor.TYPE_MAGNETIC_FIELD) != null)
        out.put("light", sm.getDefaultSensor(Sensor.TYPE_LIGHT) != null)
        out.put("proximity", sm.getDefaultSensor(Sensor.TYPE_PROXIMITY) != null)
        out.put("thermal", Build.VERSION.SDK_INT >= 29)
        call.resolve(out)
    }

    @PluginMethod
    fun start(call: PluginCall) {
        if (!running) {
            running = true
            pressureHPa = null; lux = null; proximityCm = null; mag = null
            val sm = sensorManager()
            for (type in intArrayOf(
                Sensor.TYPE_PRESSURE, Sensor.TYPE_MAGNETIC_FIELD,
                Sensor.TYPE_LIGHT, Sensor.TYPE_PROXIMITY
            )) {
                val s = sm.getDefaultSensor(type) ?: continue
                if (type == Sensor.TYPE_PROXIMITY) proximityMax = s.maximumRange
                sm.registerListener(this, s, SensorManager.SENSOR_DELAY_UI)
            }
            handler.postDelayed(emitter, 250)
        }
        call.resolve()
    }

    @PluginMethod
    fun stop(call: PluginCall) {
        teardown()
        call.resolve()
    }

    // Safety net: never keep sensors registered while the app is
    // backgrounded (the JS side also stops on visibilitychange).
    override fun handleOnPause() {
        teardown()
    }

    private fun teardown() {
        if (!running) return
        running = false
        sensorManager().unregisterListener(this)
        handler.removeCallbacks(emitter)
    }

    override fun onSensorChanged(e: SensorEvent) {
        when (e.sensor.type) {
            Sensor.TYPE_PRESSURE -> pressureHPa = e.values[0]
            Sensor.TYPE_LIGHT -> lux = e.values[0]
            Sensor.TYPE_PROXIMITY -> proximityCm = e.values[0]
            Sensor.TYPE_MAGNETIC_FIELD -> mag = e.values.clone()
        }
    }

    override fun onAccuracyChanged(sensor: Sensor?, accuracy: Int) {}

    private fun emit() {
        val out = JSObject()
        pressureHPa?.let { out.put("pressureHPa", it.toDouble()) }
        lux?.let { out.put("lux", it.toDouble()) }
        proximityCm?.let {
            out.put("proximityCm", it.toDouble())
            out.put("proximityNear", it < proximityMax)
        }
        mag?.let {
            out.put("magX", it[0].toDouble())
            out.put("magY", it[1].toDouble())
            out.put("magZ", it[2].toDouble())
        }
        val bm = context.getSystemService(Context.BATTERY_SERVICE) as BatteryManager
        val pct = bm.getIntProperty(BatteryManager.BATTERY_PROPERTY_CAPACITY)
        if (pct in 0..100) out.put("batteryPct", pct.toDouble())
        val sticky = context.registerReceiver(null, IntentFilter(Intent.ACTION_BATTERY_CHANGED))
        val status = sticky?.getIntExtra(BatteryManager.EXTRA_STATUS, -1) ?: -1
        out.put(
            "batteryCharging",
            status == BatteryManager.BATTERY_STATUS_CHARGING
                || status == BatteryManager.BATTERY_STATUS_FULL
        )
        val pm = context.getSystemService(Context.POWER_SERVICE) as PowerManager
        if (Build.VERSION.SDK_INT >= 29) {
            out.put(
                "thermal",
                when (pm.currentThermalStatus) {
                    PowerManager.THERMAL_STATUS_NONE -> "nominal"
                    PowerManager.THERMAL_STATUS_LIGHT,
                    PowerManager.THERMAL_STATUS_MODERATE -> "fair"
                    PowerManager.THERMAL_STATUS_SEVERE -> "serious"
                    else -> "critical"
                }
            )
        }
        out.put("lowPower", pm.isPowerSaveMode)
        notifyListeners("reading", out)
    }
}

File: ios-overrides/SensorProbePlugin.swift
// Capacitor plugin: stream every native-only ambient sensor the web
// layer can't reach — barometer (CMAltimeter), raw magnetometer
// (CMMotionManager), battery level/state, thermal state, Low Power
// Mode — for the Extras "Sensors" dashboard (static/extras-sensors.js).
// The Android mirror is android-overrides/SensorProbePlugin.kt; both
// register as jsName "SensorProbe" with the same contract so the JS
// side needs no platform branches:
//
//   getInfo() -> { barometer, magnetometer, light, proximity, thermal }
//       (availability booleans; light/proximity are Android-only and
//        always false here — iOS has no public API for either)
//   start()  -> begins 4 Hz 'reading' events (notifyListeners) with the
//        current snapshot: { pressureHPa?, relAltM?, magX/Y/Z?,
//        batteryPct?, batteryCharging?, thermal, lowPower }
//   stop()   -> tears the sensors down (the JS side calls this whenever
//        the dashboard is hidden, so nothing streams in the background)
//
// Notes:
//   - CMAltimeter reports pressure in kPa; converted to hPa here so the
//     JS side only ever sees hPa. relativeAltitude is metres since
//     start() (reset each session — surfaced as "altitude change").
//   - The altimeter rides the existing Motion & Fitness permission
//     (NSMotionUsageDescription is already in Info.plist for the
//     pedometer); magnetometer and the rest need no permission.
//   - UIDevice.batteryLevel needs isBatteryMonitoringEnabled = true;
//     it reports -1 until monitoring is on, which we skip over.
//
// Registered from AppBridgeViewController.capacitorDidLoad() via
// bridge?.registerPluginInstance(SensorProbePlugin()), and added to the
// Xcode project by ios-overrides/inject-into-xcodeproj.rb like the
// other override plugins.

import Foundation
import CoreMotion
import UIKit
import Capacitor

@objc(SensorProbePlugin)
public class SensorProbePlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "SensorProbePlugin"
    public let jsName = "SensorProbe"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "getInfo", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "start", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "stop", returnType: CAPPluginReturnPromise)
    ]

    private let altimeter = CMAltimeter()
    private let motion = CMMotionManager()
    private var timer: Timer?
    private var running = false
    private var pressureHPa: Double?
    private var relAltM: Double?

    @objc func getInfo(_ call: CAPPluginCall) {
        call.resolve([
            "barometer": CMAltimeter.isRelativeAltitudeAvailable(),
            "magnetometer": motion.isMagnetometerAvailable,
            "light": false,
            "proximity": false,
            "thermal": true
        ])
    }

    @objc func start(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            if self.running { call.resolve(); return }
            self.running = true
            self.pressureHPa = nil
            self.relAltM = nil
            UIDevice.current.isBatteryMonitoringEnabled = true
            if CMAltimeter.isRelativeAltitudeAvailable() {
                self.altimeter.startRelativeAltitudeUpdates(to: OperationQueue.main) { data, _ in
                    guard let d = data else { return }
                    self.pressureHPa = d.pressure.doubleValue * 10.0   // kPa -> hPa
                    self.relAltM = d.relativeAltitude.doubleValue
                }
            }
            if self.motion.isMagnetometerAvailable {
                self.motion.magnetometerUpdateInterval = 0.25
                self.motion.startMagnetometerUpdates()
            }
            self.timer = Timer.scheduledTimer(withTimeInterval: 0.25, repeats: true) { [weak self] _ in
                self?.emit()
            }
            call.resolve()
        }
    }

    @objc func stop(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            self.teardown()
            call.resolve()
        }
    }

    private func teardown() {
        guard running else { return }
        running = false
        timer?.invalidate()
        timer = nil
        altimeter.stopRelativeAltitudeUpdates()
        if motion.isMagnetometerAvailable { motion.stopMagnetometerUpdates() }
    }

    private func emit() {
        guard running else { return }
        var out = JSObject()
        if let p = pressureHPa { out["pressureHPa"] = p }
        if let a = relAltM { out["relAltM"] = a }
        if let m = motion.magnetometerData {
            out["magX"] = m.magneticField.x
            out["magY"] = m.magneticField.y
            out["magZ"] = m.magneticField.z
        }
        let level = UIDevice.current.batteryLevel
        if level >= 0 { out["batteryPct"] = Double(level) * 100.0 }
        let state = UIDevice.current.batteryState
        out["batteryCharging"] = (state == .charging || state == .full)
        switch ProcessInfo.processInfo.thermalState {
        case .nominal: out["thermal"] = "nominal"
        case .fair: out["thermal"] = "fair"
        case .serious: out["thermal"] = "serious"
        case .critical: out["thermal"] = "critical"
        @unknown default: out["thermal"] = "unknown"
        }
        out["lowPower"] = ProcessInfo.processInfo.isLowPowerModeEnabled
        notifyListeners("reading", data: out)
    }
}
